rasterio/env.py

```diff
@@ -19,7 +20,8 @@
 # Rasterio defaults
 default_options = {
     'CHECK_WITH_INVERT_PROJ': True,
-    'GTIFF_IMPLICIT_JPEG_OVR': False
+    'GTIFF_IMPLICIT_JPEG_OVR': False,
+    "I'M_ON_RASTERIO": True
```

Comment:
As long as config option keys aren't already used, there shouldn't be a problem. There's no registry of keys other than GDAL's code, though, and no guarantee of conflict. This could be a way of indicating to `osgeo.gdal` code that it is running in a `rasterio` context, but maybe I should with Even on gdal-dev about potential downsides.